lib/puppet/parser/ast/pops_bridge.rb

```diff
@@ -73,6 +73,9 @@ class Program < Puppet::Parser::AST::TopLevelConstruct
     def initialize(program_model, context = {})
       @program_model = program_model
       @context = context
+      # @todo lutter 2015-03-10: we maintain a list of definitions we
+      # instantiated so that we can attach CapabilityMappings to them
+      @types = {}
```

Comment:
Judging from the code in `Puppet::Parser::E4ParserAdapter` where the `Progam` instance is created, it seems it's created once per file which would make it impossible to have consumes/produces span over several files. A limitation that I assume is undesirable. That means the actual binding cannot be done during parsing.

Perhaps an alternative could be to have the `instantiate_CapabilityMapping` return a model object that represents the blueprint (with the name of the resource added). It could be named `CapabilityBlueprintDefinition` and inherit from `Puppet::Pops::Model::Definition`. The binding of the blueprints to their respective types could then be deferred to a new method `Puppet::Parser::Compiler::evaluate_capability_blueprints` (could be called just before `evaluate_main`). Here, the `known_resource_types` is available so it should be easy to find the referenced types.